source/windows/bcrypt_aes.c

```diff
@@ -555,120 +536,162 @@ struct aws_symmetric_cipher *aws_aes_cbc_256_new_impl(
     return NULL;
 }
 
-/* the buffer management for this mode is a good deal easier because we don't care about padding.
-   We do care about keeping the final buffer less than a block size til the finalize call so we can
-   turn the auth chaining flag off and compute the GMAC correctly. */
+/*
+ * The buffer management for this mode is a good deal easier because we don't care about padding.
+ * In chained mode, BCrypt expects the data to be passed in in multiples of block size,
+ * followed by a finalize call that turns off chaining and provides any remaining data.
+ * This function takes care of managing this state - you give it data to work and cipher state and
+ * it will return what data can be sent to bcrypt now and as side effect will update the cipher state
+ * with any leftover data.
+ * Note: this function takes a scratch buffer that might be used for to back data returned by the cursor.
+ * It is on caller to cleanup that scratch buffer.
+ */
+static struct aws_byte_cursor s_gcm_get_working_slice(
+    struct aes_bcrypt_cipher *cipher_impl,
+    struct aws_byte_cursor data,
+    struct aws_byte_buf *scratch) {
+    AWS_PRECONDITION(cipher_impl);
+    AWS_PRECONDITION(working_buf);
+
+    AWS_ZERO_STRUCT(*scratch);
+
+    struct aws_byte_cursor working_cur;
+    AWS_ZERO_STRUCT(working_cur);
+    /* If there's overflow, prepend it to the working buffer, then append the data */
+    if (cipher_impl->overflow.len) {
+        aws_byte_buf_init_copy(scratch, cipher_impl->cipher.allocator, &cipher_impl->overflow);
+        aws_byte_buf_reset(&cipher_impl->overflow, true);
+        aws_byte_buf_append_dynamic(scratch, &data);
```

Comment:
trivial: avoid the 2nd allocation, have the initial allocation be big enough
```suggestion
        struct aws_byte_cursor overflow = aws_byte_cursor_from_buf(&cipher_impl->overflow)
        aws_byte_buf_init(scratch, cipher_impl->cipher.allocator, overflow.len + data.len);
        aws_byte_buf_append(scratch, &overflow);
        aws_byte_buf_reset(&cipher_impl->overflow, true);
        aws_byte_buf_append(scratch, &data);
```

Reply:
Can this happen many times in a cipher's lifetime? Should `scratch` be a permanent struct member, to avoid re-allocating every time we need it?

Reply:
switch over to preallocating the buffer to avoid 2nd alloc. 

re lifetime: this is where things get a bit trickier, scratch will always be variable size based on overflow + input sizes.  we can have additional logic to have it ensure the size of scratch fits data for each operation. but truthfully for simplicity sake not sure its needed. and note: im not fundamentally changing how scratch space is allocated (it used to create temp buffer every time for gcm and still does for cbc and for ctr as well)